[SystemZ][z/OS] Add visibility features for z/OS (eg. _Export, pragma export)

perry-ca opened the pull request:
The z/OS operating system uses the linker to control what symbols are exported from shared libraries.  The compilation step sets a  bit on each symbol that should be exported from a shared library and then the linker marks these as exported from the shared library.  The linker also produces what is called a side deck that is a list of all the exported symbols.  This is then used as an import file during the link process for programs (or other shared libraries) that link against a shared library.  The default for compilation is to only export symbols that the user wants to export.

Control over what is exported is done by:
- making the default for visibility to be hidden
- using the _Export keyword or #pragma export to indicate what symbols should be exported.  These are features the existing compilers on z/OS use.
- using attribute((visibility,"default")) to mark the symbol as exported.

Note: `Parser::HandlePragmaExport()` has an extra layer of call to parse the pragma.  We have some more pragmas that have very similar syntax.  I'll be upstreaming those next.  It is easier to leave that layer in than to remove it for this PR and then add it back.

perry-ca committed:
initial work for pragma export & _Export keyword

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export

perry-ca committed:
Add pragma export & _Export

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export

perry-ca committed:
restore to main

perry-ca committed:
Reset pragma handler was missing

perry-ca committed:
formating

perry-ca committed:
more formating

perry-ca committed:
missing requires & s390x-none-zos targets in some

abhina-sree reviewed (approved):
LGTM

efriedma-quic commented on clang/test/CodeGen/attr-export-failing.cpp:
```diff
@@ -0,0 +1,5 @@
+// REQUIRES: systemz-registered-target
```
Do these tests actually require systemz-registered-target?

hubert-reinterpretcast replied:
Should there be diagnostics (and corresponding testing) for ignored/meaningless uses of `_Export`?

e.g.,
```cpp
typedef int _Export ty;
ty x;
int f(int _Export x);
static int _Export s;
struct S {
  int _Export nonstaticdatamember;
};
void g() {
  int _Export automatic;
}
```

AaronBallman replied:
Yes, there should be. As well as potentially confusing uses of `_Export` as in `int _Export x, y;`.

perry-ca replied:
The `int _Export x, y;` is tested in attr-export.cpp
```
int _Export var3, var4, _Export var5;
```

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -5087,6 +5087,19 @@ Decl *Sema::ParsedFreeStandingDeclSpec(Scope *S, AccessSpecifier AS,
+      if (existingValue != VisibilityAttr::Default)
+        Diag(DS.getExportSpecLoc(), diag::err_mismatched_visibility);
+    } else {
+      Tag->addAttr(
```
Not sure adding implicit attributes is necessary?

perry-ca replied:
When -fvisibility=hidden (which is the default on z/OS), etc this is needed to ensure the symbol is exported.  This is analogous to a declaration with the default visibility attribute specified.

efriedma-quic replied:
My concern here is that you have code scattered in multiple places to control the visibility; you have two different places you add a visibility attribute, and then you override the visibility in codegen.  Having overlapping ways of specifying the visibility will likely lead to weird bugs in edge cases.

perry-ca replied:
Totally valid.  The intention was to have these be another way to specify `__attribute__((visibility("default")))`.  I've rewritten the code to these issues and have hopefully addressed all of your concerns.

efriedma-quic commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -1906,6 +1906,36 @@ class Sema final : public SemaBase {
+
+  /// trySymbolLookUp try to look up a decl matching the nested specifier
+  /// with optional type list.
+  NamedDecl *trySymbolLookUp(NestedNameSpecifier *NestedName,
```
Please name this so it's clear it's specifically the lookup algorithm for pragma export.

efriedma-quic commented on clang/lib/Sema/SemaAttr.cpp:
```diff
@@ -1273,6 +1273,168 @@ void Sema::AddImplicitMSFunctionNoBuiltinAttr(FunctionDecl *FD) {
+  // Check if arguments match.
+  for (unsigned i = 0; i != FD->getNumParams(); ++i) {
+    const ParmVarDecl *PVD = FD->getParamDecl(i);
+    QualType ParmType = PVD->getOriginalType().getCanonicalType();
```
Using getOriginalType() here seems weird; does that really produce the result you want?  Please add tests with function parameters that decay (array/function types).

efriedma-quic commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -1906,6 +1906,36 @@ class Sema final : public SemaBase {
+      SymbolNames;
+  SymbolNames PendingExportNames;
+
+  FunctionDecl *tryFunctionLookUp(NestedNameSpecifier *NestedName,
```
tryFunctionLookUp is dead code?

perry-ca committed:
merge conflict

perry-ca committed:
wip

perry-ca committed:
Merge branch 'llvm:main' into perry/pragma-export

perry-ca committed:
change func names

perry-ca committed:
Merge branch 'llvm:main' into perry/pragma-export

perry-ca committed:
Merge branch 'llvm:main' into perry/pragma-export

AaronBallman commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -4456,6 +4456,12 @@ void Parser::ParseDeclarationSpecifiers(
 
+    case tok::kw__Export:
+      // If we find kw__Export, it is being applied to a var or function
+      // This will be handled in ParseDeclaratorInternal()
```
If this is a declaration specifier, shouldn't this work like any other one? e.g. https://godbolt.org/z/s4Tz1vjn4 (it's an abomination, to be sure, but it's odd to claim this as a declaration specifier but then not implement it as one).

perry-ca replied:
The _Export keyword is part of the declarator.  This code is to stop the declspec parsing.  For example, in `int _Export x,y;` only `x` is exported.  The flag was added to the DeclSpec class to handle the case of `class _Export A`.  That is later passed on to the Decl for the class.  If there is a better way to do this, I'd be interested.  We do have to match the existing XL compiler for backwards compatibility.

AaronBallman replied:
> in int _Export x,y; only x is exported.

That is deeply strange behavior IMO. I can't think of anything else in the language that works like that in that syntactic position.

> We do have to match the existing XL compiler for backwards compatibility.

That can be handled in the downstream too. For example, in this case, I would expect a diagnostic to let the user know that `_Export` very likely does not behave the way they expect in your code example.

hubert-reinterpretcast replied:
> That is deeply strange behavior IMO. I can't think of anything else in the language that works like that in that syntactic position.

In the traditional implementation, `_Export` is a declarator like `*` is, except it binds more strongly than `(<function parameters>)` or `[<array bound>]`.

AaronBallman commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1401,6 +1417,164 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
 }
 
+NestedNameSpecifier *Parser::zOSParseIdentifier(StringRef PragmaName,
+                                                IdentifierInfo *IdentName) {
```
```suggestion
                                                const IdentifierInfo *IdentName) {
```

AaronBallman commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1401,6 +1417,164 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+  }
+  while (Tok.is(tok::coloncolon)) {
+    PP.Lex(Tok);
+    IdentName = Tok.getIdentifierInfo();
```
```suggestion
    IdentifierInfo *II = Tok.getIdentifierInfo();
```
Better to use a local variable instead of reusing the parameter.

AaronBallman commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1401,6 +1417,164 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+    TypeList = SmallVector<QualType, 4>();
+    PP.Lex(Tok);
+    while (Tok.isNot(tok::eof) && !Tok.is(tok::r_paren)) {
+      //SourceRange MatchingCTypeRange;
```
Dead code?

AaronBallman commented on clang/test/CodeGen/zos-pragmas.cpp:
```diff
@@ -0,0 +1,12 @@
+// REQUIRES: systemz-registered-target
+// RUN: %clang_cc1 -x c++ -emit-llvm -triple s390x-none-zos -fvisibility=hidden %s -o - | FileCheck %s
+// expected-no-diagnostics
```
There's no `-verify` line, so this comment doesn't do anything and can be removed.

AaronBallman commented on clang/test/CodeGen/zos-pragmas.c:
```diff
@@ -0,0 +1,12 @@
+// REQUIRES: systemz-registered-target
+// RUN: %clang_cc1 -emit-llvm -triple s390x-none-zos -fvisibility=hidden %s -o - | FileCheck %s
+// expected-no-diagnostics
```
There's no `-verify` line, so this comment doesn't do anything and can be removed.

AaronBallman commented on clang/test/CodeGen/pragma-export.c:
```diff
@@ -0,0 +1,40 @@
+// REQUIRES: systemz-registered-target
+// RUN: %clang_cc1 %s -emit-llvm -fzos-extensions -triple s390x-none-zos -fvisibility=hidden -verify -o - | FileCheck %s
+
+// Testing missing declarations.
+#pragma export(d0) // expected-warning{{failed to resolve '#pragma export' to a declaration}}
+#pragma export(f9) // expected-warning{{failed to resolve '#pragma export' to a declaration}}
+
+// Testing pragma export after decl.
+void f0(void) {}
+static void sf0(void) {} // expected-warning{{#pragma export is applicable to symbols with external linkage only; not applied to 'sf0'}}
+int v0;
+static int s0; // expected-warning{{#pragma export is applicable to symbols with external linkage only; not applied to 's0'}}
+#pragma export(f0)
+#pragma export(sf0)
+#pragma export(v0)
+#pragma export(s0)
+
+// Testing pragma export before decl.
+#pragma export(f1)
+#pragma export(sf1)
+#pragma export(v1)
+#pragma export(s1)
+void f1(void) {}
+static void sf1(void) {} // expected-warning{{#pragma export is applicable to symbols with external linkage only; not applied to 'sf1'}}
+int v1;
+static int s1; // expected-warning{{#pragma export is applicable to symbols with external linkage only; not applied to 's1'}}
```
These look like they should be testing in Sema rather than in CodeGen.

AaronBallman reviewed (changes requested):
Thanks for this! Some things that are missing:

* a release note in `clang/docs/ReleaseNotes.rst`
* Parsing and Sema tests

Also, it looks like these are two separate (but related) features, one for the pragma and another for `_Export`. Is that the case? If so, they should be split into separate PRs.

perry-ca reviewed (commented):
@AaronBallman Thanks for the feedback.  I'll make the changes and look into why _Exxport wasn't implemented like the other declspecs.

I did consider these to be the same feature (two ways to set the same information).  If it's easier for reviewing I can look into splitting the PR.

perry-ca committed:
address comments

perry-ca committed:
handle arrays & func ptrs better

perry-ca committed:
Add error checking

perry-ca committed:
resolve merge conflict

perry-ca committed:
remove dead code

perry-ca committed:
formatting

perry-ca committed:
add release note

perry-ca committed:
add parsing lit tests

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export

perry-ca committed:
Improve locations in messages

perry-ca commented:
@AaronBallman @efriedma-quic.  I believe I have addressed all of your comments.  I moved the tests into Parser/Sema and added some more (and fixed problems these found).  Thanks for the feedback.

perry-ca committed:
more error handling & better reuse

perry-ca committed:
update tests

perry-ca committed:
test case updates

perry-ca committed:
merge latest

perry-ca committed:
remove unused function and improve interface to other

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export

perry-ca committed:
handle externC properly for C++

AaronBallman reviewed (approved):
I didn't spot any further concerns, but I'd appreciate a second set of eyes on this given the size of the changes.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7552,6 +7552,27 @@ attribute requires a string literal argument to identify the handle being releas
+  let Content = [{
+Use the ``_Export`` keyword on a function or external variable to declare
+that it is to be exported (made available to other modules). You must define
+the object name in the same translation unit in which you use the ``_Export``
```
This documentation should also mention the pragma spelling and document the whole thing.  Also, do we diagnose that `anothony` isn't defined?

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7552,6 +7552,27 @@ attribute requires a string literal argument to identify the handle being releas
+
+This statement exports the function ``anthony``, if you define the function in the
+translation unit. The ``_Export`` keyword must immediately precede the declaration name.
+If you apply the ``_Export`` keyword to a class, the compiler automatically exports
```
```suggestion
If you apply the ``_Export`` keyword to a class or struct, the compiler automatically exports
```

What does it do with unions?

erichkeane commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -4504,6 +4504,11 @@ void Parser::ParseDeclarationSpecifiers(
 
+    case tok::kw__Export:
+      // We're done with the declaration-specifiers.
+      goto DoneWithDeclSpec;
```
Does this mean `_Export` has to be last?   Not a parser guy here, but this one doesn't feel right to me?  @aaronballman can you double check here?

hubert-reinterpretcast replied:
`_Export` is not really a declaration specifier (at least in the traditional implementation).
It is more like a C++11 attribute except that it lives on the left of a _declarator-id_ instead of on the right.
In some realizations, the (less clean) conceptual model is that it is a declarator like `*` is, except it binds more strongly than `(<function parameters>)` or `[<array bound>]`.

@perry-ca, I don't see any test in this PR for `_Export` in positions like:
```cpp
int (*_Export x)(void) = 0;
```

Is the above accepted with this PR? Which conceptual model is being implemented here?

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -401,6 +401,12 @@ struct PragmaMaxTokensTotalHandler : public PragmaHandler {
                     Token &FirstToken) override;
 };
 
+struct PragmaExportHandler : public PragmaHandler {
```
This patch is big enough... I am beginning to wonder if we should split this pragma off into a separate patch.

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1400,6 +1416,171 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+NestedNameSpecifier *
+Parser::zOSParseIdentifier(StringRef PragmaName,
+                           const IdentifierInfo *IdentName) {
+  NestedNameSpecifier *NestedId = nullptr;
```
What is this for the grammar of?  Are you just trying to get the name of a thing?  I did something similar, see : `ParseOpenACCIDExpression`.  For C++ you can use `ParseCXXIdExpression`, and I emulated it for C.

erichkeane reviewed (commented):
There seems to be a LOT of added complexity for the `#pragma export`, to the point that I'm not sure of the value compared to the keyword, and if we're going to keep it, we need to spend quite a bit more time on it.

As far as the keyword itself, I think there are quite a few more diagnostics we should look at, particularly for 'completeness'.  I don't really like the error that is being diagnosed (just 'cannot be qualified'), as that is really quite vague.  We need to try harder on that.
